src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
         /// <returns>The number of decoded symbols.</returns>
         public static int Decode(ReadOnlySpan<byte> src, ref byte[] dstArray)
         {
+            // The code below implements the decoding logic for an HPack huffman encoded literal values.
+            // https://httpwg.org/specs/rfc7541.html#string.literal.representation
+            //
+            // To decode a symbol, we traverse the decoding lookup table tree by 8 bits for each lookup
+            // until we found a leaf - which contains decoded symbol (octet)
+            //
+            // see comments in GenerateDecodingLookupTree() describing decoding table
+
             Span<byte> dst = dstArray;
             Debug.Assert(dst != null && dst.Length > 0);
 
+            if (s_decodingTree == null)
+            {
+                Volatile.Write(ref s_decodingTree, GenerateDecodingLookupTree());
+            }
+            ushort[] decodingTree = s_decodingTree;
+
+            int lookupTableIndex = 0;
+            int lookupIndex;
+
+            ushort acc = 0;
+            int bitsInAcc = 0;
+
+
             int i = 0;
             int j = 0;
-            int lastDecodedBits = 0;
             while (i < src.Length)
```

Comment:
From the assert in L414 we know that there will be at least one element, so could change this to da `do-while`-loop, thus eliminating one length-check before loop entry.

Reply:
L414 asserts dst not src. Expecting src has at least one byte would break contract of previous implementation.

Reply:
Oh, sorry...